test/absinthe/schema/sdl_render_test.exs

```diff
@@ -0,0 +1,283 @@
+
+    assert rendered == TestSchema.sdl()
+  end
+
+  @builtin ["String", "Int", "Float", "Boolean", "ID"]
+
+  def render(%{"name" => "__" <> _introspection_type}) do
+    empty()
+  end
+
+  def render(%{"ofType" => nil, "kind" => "SCALAR", "name" => name}) do
+    name
+  end
+
+  def render(%{"ofType" => nil, "name" => name}) do
+    name
+  end
+
+  def render(%{"ofType" => type, "kind" => "LIST"}) do
+    concat(["[", render(type), "]"])
+  end
+
+  def render(%{"ofType" => type, "kind" => "NON_NULL"}) do
+    concat([render(type), "!"])
+  end
+
+  def render(%{"kind" => "SCALAR", "name" => name} = thing) when name in @builtin do
+    empty()
+  end
+
+  def render(%{
+        "defaultValue" => _,
+        "name" => name,
+        "description" => description,
+        "type" => arg_type
+      }) do
+    maybe_description(
+      description,
+      concat([
+        name,
+        ": ",
+        render(arg_type)
+      ])
+    )
+  end
+
+  def render(%{
+        "name" => name,
+        "args" => args,
+        "type" => field_type
+      }) do
+    arg_docs = Enum.map(args, &render/1)
+
+    concat([
+      name,
+      maybe_args(arg_docs),
+      ": ",
+      render(field_type)
+    ])
+  end
+
+  def render(%{
+        "kind" => "OBJECT",
+        "name" => name,
+        "description" => description,
+        "fields" => fields
+      }) do
+    field_docs = Enum.map(fields, &render/1)
+
+    maybe_description(
+      description,
+      block(
+        "type",
+        name,
+        join_with(field_docs, line())
+      )
+    )
+  end
+
+  def render(%{
+        "kind" => "UNION",
+        "name" => name,
+        "description" => description,
+        "possibleTypes" => possible_types
+      }) do
+    possible_type_docs = Enum.map(possible_types, & &1["name"])
+
+    maybe_description(
+      description,
+      concat([
+        "union",
+        " ",
+        name,
+        " = ",
+        join_with(possible_type_docs, " | ")
+      ])
+    )
+  end
+
+  def render(%{
+        "kind" => "ENUM",
+        "name" => name,
+        "description" => description,
+        "enumValues" => values
+      }) do
+    value_names = Enum.map(values, & &1["name"])
+
+    maybe_description(
+      description,
+      block(
+        "enum",
+        name,
+        join_with(value_names, line())
+      )
+    )
+  end
+
+  def render(type) do
+    IO.inspect(type, label: "MISSIN")
+    empty()
+  end
+
+  def maybe_description(nil, docs), do: docs
+
+  def maybe_description(description, docs) do
+    join_with(
+      if String.contains?(description, "\n") do
+        [
+          join_with(["\"\"\"", description, "\"\"\""], line()),
+          docs
+        ]
+      else
+        [
+          concat(["\"", description, "\""]),
+          docs
+        ]
+      end,
+      line()
+    )
+  end
+
+  def maybe_args([]) do
+    empty()
+  end
+
+  def maybe_args(docs) do
+    # TODO:
+    #  figure out 1 line vs multi-line args
```

Comment:
I need to understand how `Inspect.Algebra` works better so that args will split to multi-line in the correct cases, time to read more of the Elixir code formatter code!